platform/src/SaveablePanel.js

```diff
@@ -0,0 +1,134 @@
+import { Panel } from "./Panel.js";
+import * as Diff from 'diff';
+
+class SaveablePanel extends Panel {
+    fileUrl;
+    valueSha;
+    lastSavedContent;
+    diff;
+
+    constructor(id) {
+        super(id);
+    }
+
+    initialize(editor) {
+        super.initialize(editor);
+
+        // Add an event listener to the editor to update the diff on change
+        this.editor.session.on("change", () => {
+            this.updatePanelDiff();
+        })
+    }
+
+    /**
+     * Updates the internal diff state of the panel.
+     * 
+     * Compares the current panel content with the last saved content
+     * and stores the resulting differences in the `diff` property.
+     * 
+     * This method is typically triggered whenever the editor content changes.
+     * The computed diff is a filtered list of added or removed lines,
+     * ignoring unchanged lines.
+     * 
+     * Uses jsdiff's `diffLines` to compute the differences.
+     */
+    updatePanelDiff() {
+
+        const currentValue = this.getValue() ?? "";
+        const lastSavedValue = this.getLastSavedContent() ?? "";
+
+        if (currentValue === lastSavedValue) {
+            this.diff = []; // No changes, set diff to empty array
+            return;
+        }
+
+        this.diff = Diff.diffLines(lastSavedValue, currentValue)
+            .map(part => {
+                let change = {};
+                if (part.added) {
+                    change.added = part.value;
+                }
+                else if (part.removed) {
+                    change.removed = part.value;
+                }
+                return change;
+            })
+            .filter(change => change.added || change.removed);
+    }
+
+    getDiff() {
+        return this.diff;
+    }
+
+    getFileUrl() {
+        return this.fileUrl;
+    }
+
+    setFileUrl(url) {
+        this.fileUrl = url;
+    }
+
+    /**
+     * Currently only supports GitHub URLs.
```

Comment:
```suggestion
     * FIXME: Currently only supports GitHub URLs.
```
I guess we would want to generalise this at some point (though not right now...)

Reply:
I think this is something that can be moved alongside the Strategy pattern with the Filehandler and VCS provider files. Instead of doing `panel.getFilePath()`, we could add a method in the FileHandler called `getFilePath(fileUrl)`. This would lead to `findProvider`, which would choose the VCS provider at runtime, and each VCS provider can have its own implementation of finding a file path from a given URL.

Reply:
> I think this is something that can be moved alongside the Strategy pattern with the Filehandler and VCS provider files. Instead of doing `panel.getFilePath()`, we could add a method in the FileHandler called `getFilePath(fileUrl)`. This would lead to `findProvider`, which would choose the VCS provider at runtime, and each VCS provider can have its own implementation of finding a file path from a given URL.

Implemented in [9573744](https://github.com/mdenet/educationplatform/pull/227/commits/957374455390776c68e187bccb3a47dbfa5ce336)

Each VCS provider has its own structure for a raw URL, so we delegate this to its respective provider class.